### PR TITLE
Implementation of SOAP plugin slot

### PR DESCRIPTION
**Pull request by wanze**
**Feature-Wiki**
http://www.ilias.de/docu/goto.php?target=wiki_1357_SOAP_Plugin_Slot

**Example Plugin**
https://github.com/ILIAS-eLearning/TestSoapHook

### CODE REVIEW

**Comment by chfsx** [comment]
Hi @smeyer-ilias this PR represents the accepted feature in http://www.ilias.de/docu/goto.php?target=wiki_1357_SOAP_Plugin_Slot. Could you please merge it to trunk? Thanks a lot!

File: webservice/soap/include/inc.soap_functions.php [hunk 1]
```diff
@@ -1063,6 +1067,37 @@ public static function processBackgroundTask($sid, $task_id) {
...
+	 */
+	public function __call($name, $arguments) {
+		// SoapHookPlugins need the client-ID submitted
+		if (!isset($_GET['client_id'])) {
+			throw new SoapFault('SOAP-ENV:Server', "Function '{$name}' does not exist");
+		}
+		// Note: We need to bootstrap ILIAS in order to get $ilPluginAdmin and load the soap plugins.
+		// We MUST use a context that does not handle authentication at this point (session is checked by SOAP).
+		// Solution: Use the REST context
+		ilContext::init(ilContext::CONTEXT_REST);
```
Comment (smeyer-ilias) [inline webservice/soap/include/inc.soap_functions.php]: Please can you create a new context for the first initialisation. Reusing a different context is to error prone.

**Review by smeyer-ilias — COMMENTED** [review]
Thanks for this PR. I will check the plugin on Tuesday and give a final feedback.

File: webservice/soap/include/inc.soap_functions.php (hunk 1 above)
Comment (wanze) [reply webservice/soap/include/inc.soap_functions.php]: @smeyer-ilias What would be the preferred name for such context? The context needs to init the client but not handle authentication. Something like `ilContextSoapNoAuth` ?

Comment (smeyer-ilias) [reply webservice/soap/include/inc.soap_functions.php]: Yes, this would be ok.

**Comment by chfsx** [comment]
Hi @smeyer-ilias did you already had the time for a review? If there are any questions feel free to contact me directly. Best regards, Fabian

File: webservice/soap/classes/class.ilNusoapUserAdministrationAdapter.php [hunk 2]
```diff
@@ -1350,10 +1352,50 @@ function __registerMethods()
...
 
 	}
 
+	/**
+	 * Register any methods and types of SOAP plugins to the SOAP server
+	 */
+	protected function handleSoapPlugins() {
+		// Note: We need a context that does not handle authentication at this point, because this is
+		// handled by an actual SOAP request which always contains the session ID and client
+		ilContext::init(ilContext::CONTEXT_REST);
```
Comment (smeyer-ilias) [inline webservice/soap/classes/class.ilNusoapUserAdministrationAdapter.php]: The new context ilContext::CONTEXT_SOAP_NO_AUTH could be used here too.

File: Services/WebServices/SOAP/classes/class.ilAbstractSoapMethod.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	 * which are both required in order to handle the request.
+	 *
+	 * @param string $session_id
+	 * @throws ilSoapPluginException
+	 */
+	protected function initIliasAndCheckSession($session_id) {
+		$this->initAuth($session_id);
+		$this->initIlias();
+		if (!$this->__checkSession($session_id)) {
+			throw new ilSoapPluginException('Invalid session-ID provided');
```
Comment (smeyer-ilias) [inline Services/WebServices/SOAP/classes/class.ilAbstractSoapMethod.php]: throw new ilSoapPluginException($this->__getMessage()); will keep the original error message from __checkSession

**Review by smeyer-ilias — COMMENTED** [review]
As mentioned by call the WSDL does not return the correct "address location" if called with parameter "client_id". I tried something like :
	if(isset($_GET['client_id'])) $attrs['location'] .= '?client_id='.$_GET['client_id'];
in file webservice/soap/lib/nusoap.php line 5495.

File: webservice/soap/server.php [hunk 4]
```diff
@@ -39,15 +38,18 @@
...
-	// called by webservice
-	//ini_set("soap.wsdl_cache_enabled", "1"); 
+	// This is a SOAP request
 	include_once('webservice/soap/include/inc.soap_functions.php');
-	$soapServer = new SoapServer(ilSoapFunctions::buildHTTPPath()."/webservice/soap/nusoapserver.php?wsdl");
-	$soapServer->setClass("ilSoapFunctions");
-	$soapServer->handle();				
-} 
+	$uri = ilSoapFunctions::buildHTTPPath() . '/webservice/soap/server.php';
+	if (isset($_GET['ref_id'])) {
```
Comment (smeyer-ilias) [inline webservice/soap/server.php]: Really? Shouldn't it be $_GET['client_id'] ?

Comment (chfsx) [reply webservice/soap/server.php]: of course, thanks!

File: Services/WebServices/SOAP/classes/class.ilAbstractSoapMethod.php (hunk 3 above)
Comment (chfsx) [reply Services/WebServices/SOAP/classes/class.ilAbstractSoapMethod.php]: good hint, I changed it accordingly

File: webservice/soap/classes/class.ilNusoapUserAdministrationAdapter.php (hunk 2 above)
Comment (chfsx) [reply webservice/soap/classes/class.ilNusoapUserAdministrationAdapter.php]: thanks! changed it accordingly!

**Comment by chfsx** [comment]
Hi @smeyer-ilias, thank you very much for the review, I changed the code according to your recommendations! Best regards, Fabian

**Comment by smeyer-ilias** [comment]
Thanks Stefan, thanks Fabian.

**Pull request merged**